Buf LSP panics in DocumentSymbol via ir.Type.Deprecated on devel build

### GitHub repository with your minimal reproducible example (do not leave this field blank or fill out this field with `github.com/bufbuild/buf` or we will automatically close your issue, see the instructions above!)

Private repository. I can try to reduce this to a minimal public repro if needed, but I wanted to report the exact crash signature first because it looks like a new variant of earlier Buf LSP panics around deprecated-symbol handling.

### Commands

```shell
buf lsp serve --timeout=0 --log-format=json
```

### Output

```text
INFO	starting LSP server	{"buf_version":"(devel)"}
INFO	client attached	{"buf_version":"(devel)","client_name":"Visual Studio Code","client_version":"1.114.0"}
INFO	file updated	{"buf_version":"(devel)","uri":"/Users/mklein/Source/internal-core/proto-internal/src/bitdrift/internal_api/unary/explorations/v1/workflow.proto","old_version":-1,"new_version":1}
INFO	parsing IR for file	{"buf_version":"(devel)","uri":"file:///Users/mklein/Source/internal-core/proto-internal/src/bitdrift/internal_api/unary/explorations/v1/workflow.proto","version":1}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x103e3f40c]

goroutine 31 [running]:
github.com/bufbuild/protocompile/experimental/ir.MessageValue.Field({{{0x0?}, 0x0?, 0x44d6345?}}, {{{0x0?}, 0x0?, 0x0?}})
  github.com/bufbuild/protocompile@v0.14.2-0.20260306221011-519528254156/experimental/ir/ir_value.go:799 +0xcc
github.com/bufbuild/protocompile/experimental/ir.Type.Deprecated({{{0x5444046a38c8?}, 0x544404bc0008?, 0x361?}})
  github.com/bufbuild/protocompile@v0.14.2-0.20260306221011-519528254156/experimental/ir/ir_type.go:596 +0xa4
github.com/bufbuild/protocompile/experimental/ir.Symbol.Deprecated({{{0x5444046a38c8?}, 0x544406baa60c?, 0x4140518?}})
  github.com/bufbuild/protocompile@v0.14.2-0.20260306221011-519528254156/experimental/ir/ir_symbol.go:140 +0x70
github.com/bufbuild/buf/private/buf/buflsp.(*symbol).GetSymbolInformation(0x544405f85e60)
  github.com/bufbuild/buf/private/buf/buflsp/symbol.go:359 +0x14c
github.com/bufbuild/buf/private/buf/buflsp.(*server).DocumentSymbol.(*file).GetSymbols.func1(...)
  github.com/bufbuild/buf/private/buf/buflsp/file.go:1206
github.com/bufbuild/buf/private/buf/buflsp.(*server).DocumentSymbol(0x544403a68cf0?, {0x105686840?, 0x5444058dc760?}, 0x103de50c8?)
  github.com/bufbuild/buf/private/buf/buflsp/server.go:500 +0x168
go.lsp.dev/protocol.serverDispatch({0x105c7dfb0, 0x544404066230}, {0x105ce53b8, 0x54440441aa68}, 0x54440496bea8, {0x105c7fc38, 0x5444040514c0})
  go.lsp.dev/protocol@v0.12.0/server.go:386 +0x6e94
github.com/bufbuild/buf/private/buf/buflsp.(*lsp).newHandler.ServerHandler.func2({0x105c7dfb0, 0x544404066230}, 0x54440496bea8, {0x105c7fc38, 0x5444040514c0})
  go.lsp.dev/protocol@v0.12.0/server.go:36 +0x68
github.com/bufbuild/buf/private/buf/buflsp.(*lsp).newHandler.func1({0x105c7dfb0, 0x544404066230}, 0x54440441ac90, {0x105c7fc38, 0x5444040514c0})
  github.com/bufbuild/buf/private/buf/buflsp/buflsp.go:190 +0x290
github.com/bufbuild/buf/private/buf/buflsp.(*lsp).newHandler.AsyncHandler.func3.2()
  go.lsp.dev/jsonrpc2@v0.10.0/handler.go:114 +0x6c
created by github.com/bufbuild/buf/private/buf/buflsp.(*lsp).newHandler.AsyncHandler.func3 in goroutine 27
  go.lsp.dev/jsonrpc2@v0.10.0/handler.go:112 +0x140
[Error - 9:01:01 AM] The Buf Language Server server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
[Error - 9:01:01 AM] Server process exited with code 2.
```

### Expected Output

The LSP should not panic when serving `textDocument/documentSymbol` for this file.

### Anything else?

Environment:
- Buf version: `(devel)`
- VS Code: `1.114.0`
- OS: macOS
- protocompile in stack: `v0.14.2-0.20260306221011-519528254156`

This looks related to earlier Buf LSP crash reports, but I did not find an exact duplicate for this stack:
- #4192
- #4218
- #4234

What seems distinct here is that the panic goes through:
- `ir.MessageValue.Field`
- `ir.Type.Deprecated`
- `ir.Symbol.Deprecated`
- `buflsp.(*symbol).GetSymbolInformation`
- `DocumentSymbol`

So this looks like another nil-deref in deprecated-symbol handling, but for `Type.Deprecated` on the current protocompile snapshot rather than the previously reported cases.

I have not yet reduced this to a minimal public repro. If it helps triage, I can try to isolate the specific proto shape or import graph that triggers it.

--- a/go.mod
+++ b/go.mod
@@ -17,7 +17,7 @@ require (
 	connectrpc.com/connect v1.19.1
 	connectrpc.com/grpcreflect v1.3.0
 	connectrpc.com/otelconnect v0.9.0
-	github.com/bufbuild/protocompile v0.14.2-0.20260406184405-b06501d51312
+	github.com/bufbuild/protocompile v0.14.2-0.20260409171342-5665f7ecc74e
 	github.com/bufbuild/protoplugin v0.0.0-20250218205857-750e09ce93e1
 	github.com/cli/browser v1.3.0
 	github.com/gofrs/flock v0.13.0

--- a/go.mod
+++ b/go.mod
@@ -17,7 +17,7 @@ require (
 	connectrpc.com/connect v1.19.1
 	connectrpc.com/grpcreflect v1.3.0
 	connectrpc.com/otelconnect v0.9.0
-	github.com/bufbuild/protocompile v0.14.2-0.20260409171342-5665f7ecc74e
+	github.com/bufbuild/protocompile v0.14.2-0.20260409205102-f58612be4b0a
 	github.com/bufbuild/protoplugin v0.0.0-20250218205857-750e09ce93e1
 	github.com/cli/browser v1.3.0
 	github.com/gofrs/flock v0.13.0

--- a/private/buf/buflsp/deprecate.go
+++ b/private/buf/buflsp/deprecate.go
@@ -42,8 +42,8 @@ func (s *server) getDeprecateCodeAction(
 		slog.Uint64("line", uint64(params.Range.Start.Line)),
 		slog.Uint64("char", uint64(params.Range.Start.Character)),
 	)
-	if file.workspace == nil || file.ir == nil {
-		s.logger.DebugContext(ctx, "deprecate: no workspace or IR")
+	if file.workspace == nil || file.ir == nil || !file.ir.Lowered() {
+		s.logger.DebugContext(ctx, "deprecate: no workspace or valid IR")
 		return nil
 	}
 
@@ -62,7 +62,7 @@ func (s *server) getDeprecateCodeAction(
 	checker := newFullNameMatcher(fqnPrefix)
 	edits := make(map[protocol.DocumentURI][]protocol.TextEdit)
 	for _, wsFile := range file.workspace.PathToFile() {
-		if wsFile.ir == nil {
+		if wsFile.ir == nil || !wsFile.ir.Lowered() {
 			continue
 		}
 		fileEdits := generateDeprecationEdits(wsFile, checker)
@@ -191,7 +191,7 @@ func generateDeprecationEdits(file *file, checker *fullNameMatcher) []protocol.T
 
 // getPackageFQN extracts the package FQN components from a file.
 func getPackageFQN(file *file) ir.FullName {
-	if file.ir == nil {
+	if file.ir == nil || !file.ir.Lowered() {
 		return ""
 	}
 	return file.ir.Package()

--- a/private/buf/buflsp/diagnostic.go
+++ b/private/buf/buflsp/diagnostic.go
@@ -17,6 +17,8 @@
 package buflsp
 
 import (
+	"strings"
+
 	"github.com/bufbuild/protocompile/experimental/report"
 	"github.com/bufbuild/protocompile/experimental/report/rtags"
 	"github.com/bufbuild/protocompile/experimental/source/length"
@@ -41,10 +43,16 @@ var reportLevelToDiagnosticSeverity = map[report.Level]protocol.DiagnosticSeveri
 func reportDiagnosticToProtocolDiagnostic(
 	reportDiagnostic report.Diagnostic,
 ) protocol.Diagnostic {
+	message := reportDiagnostic.Message()
+	if reportDiagnostic.Level() == report.ICE {
+		// Include notes for ICE
+		notes := append([]string{message}, reportDiagnostic.Notes()...)
+		message = strings.Join(notes, " ")
+	}
 	diagnostic := protocol.Diagnostic{
 		Source:   serverName,
 		Severity: reportLevelToDiagnosticSeverity[reportDiagnostic.Level()],
-		Message:  reportDiagnostic.Message(),
+		Message:  message,
 	}
 	if primary := reportDiagnostic.Primary(); !primary.IsZero() {
 		startLocation := primary.Location(primary.Start, positionalEncoding)

--- a/private/buf/buflsp/diagnostics_test.go
+++ b/private/buf/buflsp/diagnostics_test.go
@@ -192,11 +192,11 @@ func TestDiagnostics(t *testing.T) {
 				{
 					Range: protocol.Range{
 						Start: protocol.Position{Line: 8, Character: 0},
-						End:   protocol.Position{Line: 8, Character: 0},
+						End:   protocol.Position{Line: 8, Character: 1},
 					},
 					Severity: protocol.DiagnosticSeverityError,
 					Source:   "buf-lsp",
-					Message:  "syntax error: expecting ';'",
+					Message:  "unexpected `}` after definition",
 				},
 			},
 		},
@@ -226,12 +226,12 @@ func TestDiagnostics(t *testing.T) {
 			expectedDiagnostics: []protocol.Diagnostic{
 				{
 					Range: protocol.Range{
-						Start: protocol.Position{Line: 5, Character: 7},
-						End:   protocol.Position{Line: 5, Character: 28},
+						Start: protocol.Position{Line: 5, Character: 0},
+						End:   protocol.Position{Line: 5, Character: 30},
 					},
 					Severity: protocol.DiagnosticSeverityError,
 					Source:   "buf-lsp",
-					Message:  `does/not/exist.proto: file does not exist`,
+					Message:  "imported file does not exist",
 				},
 			},
 		},

--- a/private/buf/buflsp/file.go
+++ b/private/buf/buflsp/file.go
@@ -57,6 +57,8 @@ type file struct {
 	// diagnostics or symbols an operating refers to.
 	version int32
 	hasText bool // Whether this file has ever had text read into it.
+	// The local path of the descriptor.proto file used for compilation. This is stored for diagnostics.
+	descriptorProtoLocalPath string
 
 	workspace  *workspace         // May be nil.
 	objectInfo storage.ObjectInfo // Info in the context of the workspace.
@@ -240,7 +242,7 @@ func (f *file) RefreshIR(ctx context.Context) {
 	for path, file := range pathToFiles {
 		current := openerMap[path]
 		// If there is no entry for the current path or if the file content has changed, we
-		// update the opener and set a new query.
+		// update the opener and evict stale [queries.File] queries.
 		if current == nil || current.Text() != file.file.Text() {
 			openerMap[path] = file.file
 			if current != nil {
@@ -250,6 +252,7 @@ func (f *file) RefreshIR(ctx context.Context) {
 		}
 		files = append(files, file)
 	}
+
 	// Remove paths that are no longer in the current workspace and evict stale query keys.
 	for path := range openerMap {
 		if _, ok := pathToFiles[path]; !ok {
@@ -276,21 +279,30 @@ func (f *file) RefreshIR(ctx context.Context) {
 		)
 		return
 	}
+
 	for i, file := range files {
 		file.ir = results[i].Value
+		if file.objectInfo.Path() == "google/protobuf/descriptor.proto" {
+			// Set local path for the descriptor.proto file used for diagnostics.
+			f.descriptorProtoLocalPath = file.objectInfo.LocalPath()
+		}
 		if f != file {
 			// Update symbols for imports.
 			file.IndexSymbols(ctx)
 		}
 	}
+
 	// Store the IR report for code actions
 	f.irReport = diagnosticReport
 
-	// Only hold on to diagnostics where the primary span is for this path.
+	// Only hold on to diagnostics where the primary span is for this file.
 	fileDiagnostics := xslices.Filter(diagnosticReport.Diagnostics, func(d report.Diagnostic) bool {
-		return d.Primary().Path() == f.objectInfo.Path()
+		// We avoid returning warnings from the compiler for now, since these may conflate with
+		// lint checks.
+		return d.Primary().Path() == f.objectInfo.LocalPath() && d.Level() < report.Warning
 	})
 	f.diagnostics = xslices.Map(fileDiagnostics, reportDiagnosticToProtocolDiagnostic)
+
 	f.lsp.logger.DebugContext(
 		ctx, "ir diagnostic(s)",
 		slog.String("uri", f.uri.Filename()),
@@ -338,8 +350,23 @@ func (f *file) queryFileKeys() []any {
 // This operation requires RefreshIR.
 func (f *file) IndexSymbols(ctx context.Context) {
 	defer xslog.DebugProfile(f.lsp.logger, slog.String("uri", string(f.uri)))()
-	// We cannot index symbols without the IR, so we keep the symbols as-is.
+
 	if f.ir == nil {
+		return
+	}
+
+	// The file has not completed lowering, so we cannot continue with symbol indexing here.
+	// To help the user better understand/diagnose this problem, we surface an additional
+	// diagnostic here.
+	if !f.ir.Lowered() {
+		f.diagnostics = append(f.diagnostics, protocol.Diagnostic{
+			Source:   serverName,
+			Severity: protocol.DiagnosticSeverityError,
+			Message: fmt.Sprintf(`the symbols for this file have not been fully resolved due to an invalid version of descriptor.proto located at: %q.
+this is likely due to a vendored descriptor.proto.`,
+				f.descriptorProtoLocalPath,
+			),
+		})
 		return
 	}
 

--- a/private/buf/buflsp/file.go
+++ b/private/buf/buflsp/file.go
@@ -362,8 +362,8 @@ func (f *file) IndexSymbols(ctx context.Context) {
 		f.diagnostics = append(f.diagnostics, protocol.Diagnostic{
 			Source:   serverName,
 			Severity: protocol.DiagnosticSeverityError,
-			Message: fmt.Sprintf(`the symbols for this file have not been fully resolved due to an invalid version of descriptor.proto located at: %q.
-this is likely due to a vendored descriptor.proto.`,
+			Message: fmt.Sprintf(`The symbols for this file have not been fully resolved due to an invalid version of descriptor.proto located at: %q.
+This is likely due to a vendored descriptor.proto.`,
 				f.descriptorProtoLocalPath,
 			),
 		})
@@ -921,7 +921,7 @@ func (f *file) messageToSymbolsHelper(msg ir.MessageValue, index int, parents []
 		// each path component.
 		for element := range seq.Values(field.Elements()) {
 			key := field.KeyASTs().At(element.ValueNodeIndex())
-			components := slices.Collect(key.AsPath().Components)
+			components := slices.Collect(key.AsPath().Components())
 			// If there are no path components for an element, then we skip it, since there are
 			// no symbols to track.
 			if len(components) == 0 {

--- a/private/buf/buflsp/folding_range.go
+++ b/private/buf/buflsp/folding_range.go
@@ -26,7 +26,7 @@ import (
 // foldingRange generates folding ranges for a file.
 // It finds messages, services, enums, and multi-line comment blocks.
 func (s *server) foldingRange(file *file) []protocol.FoldingRange {
-	if file.ir == nil {
+	if file.ir == nil || !file.ir.Lowered() {
 		return nil
 	}
 

